File: cpp/98.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return valid(root, NULL, NULL);
    }
    
    bool valid(TreeNode *root, TreeNode *minNode, TreeNode *maxNode){
        if(!root) return true;
        if(minNode && root->val <= minNode->val ||maxNode && root->val >= maxNode->val ) return false;
        return valid(root->left, minNode, root) && valid(root->right, root, maxNode);
    }
};

File: golang/98/98.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	return helper(root, nil, nil)
}

func helper(root *TreeNode, min, max *int) bool {
	if root == nil {
		return true
	}

	if min != nil && root.Val <= *min {
		return false
	}

	if max != nil && root.Val >= *max {
		return false
	}

	return helper(root.Left, min, &root.Val) && helper(root.Right, &root.Val, max)
}
